### PR TITLE
 feat(vm-snapshots): add ability to create, see list of vm snapshots in the vm sidebar

### PR DESCRIPTION
**Pull request by tamazlykar**
Implements #928

### CODE REVIEW

File: src/app/root-store/server-data/vm-snapshots/vm-snapshots.effects.ts [hunk 1]
```diff
@@ -57,7 +63,59 @@ export class VmSnapshotsEffects {
...
+      );
+      const params = {
+        virtualmachineid: payload.vmId,
+        description: payload.description,
+        name: payload.name,
+        snapshotmemory: payload.snapshotMemory,
+      };
+      return this.vmSnapshotsService.create(params).pipe(
+        tap(() => {
+          const message = 'NOTIFICATIONS.VM_SNAPSHOTS.TAKE_VM_SNAP_IN_DONE';
```
Comment (wowshakhov) [inline src/app/root-store/server-data/vm-snapshots/vm-snapshots.effects.ts]: IN_DONE? Do you mean IS_DONE?

File: src/app/root-store/server-data/vm-snapshots/vm-snapshots.state.ts [hunk 2]
```diff
@@ -1,12 +1,21 @@
 import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';
 import { VmSnapshot } from '../../../shared/models/vm-snapshot.model';
+import moment = require('moment');
 
 export interface VmSnapshotsState extends EntityState<VmSnapshot> {
   isLoading: boolean;
 }
 
+const sortByCreationDate = (a: VmSnapshot, b: VmSnapshot) => {
+  const aDate = moment(a.created).toDate();
```
Comment (wowshakhov) [inline src/app/root-store/server-data/vm-snapshots/vm-snapshots.state.ts]: replace with bTimestamp.localeCompare(aTimestamp)

File: src/i18n/en.json [hunk 3]
```diff
@@ -585,6 +595,20 @@
       "NO_OFFERINGS": "No offerings available",
       "GENERAL": "General",
       "ADVANCED": "Advanced"
+    },
+    "MODALS": {
+      "TAKE_A_VM_SNAPSHOT": {
+        "TITLE": "Take VM Snapshot",
+        "SNAPSHOT_MEMORY": "Snapshot memory",
+        "WARNING_INFORMATION": "Please notice that the instance will be paused during the snapshoting, and resumed after snapshotting."
```
[the remark this answers is not in the document]
Comment (tamazlykar) [reply src/i18n/en.json]: This message was taken from the original interface.

File: src/app/root-store/server-data/vm-snapshots/vm-snapshots-entity-actions.selectors.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      ];
+      return dictionary;
+    }, {});
+  },
+);
+
+export const getVmSnapshotEntityActions = (vmSnapshotId: string) =>
+  createSelector(
+    getVmSnapshotEntityActionsEntities,
+    actionsEntities => actionsEntities[vmSnapshotId] || [],
```
[the remark this answers is not in the document]
Comment (tamazlykar) [reply src/app/root-store/server-data/vm-snapshots/vm-snapshots-entity-actions.selectors.ts]: The word "entities" is used throughout the project. This is a standard name of the selector "selectEntities" from Ngrx Entities library which means the map of entities by ID

File: src/app/root-store/server-data/vm-snapshots/vm-snapshots-entity-actions.selectors.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      const vm = virtualMachineEntities[vmSnapshot.virtualmachineid];
+      const vmState = vm ? vm.state : undefined;
+      dictionary[vmSnapshot.id] = [
+        {
+          icon: 'mdi-camera',
+          text: 'SNAPSHOT_PAGE.ACTIONS.CREATE_SNAP_FROM_VM_SNAP',
+          disabled: false,
+          visible: false,
+          actionCreator: () =>
+            new vmSnapshotsActions.CreateVolumeSnapshot({ snapshotId: vmSnapshot.id }),
```
Comment (wowshakhov) [inline src/app/root-store/server-data/vm-snapshots/vm-snapshots-entity-actions.selectors.ts]: The other two actions are named Revert and Delete. Is there any reason to have `CreateVolumeSnapshot` instead of `Create`?

File: src/app/shared/components/volume-snapshot-from-vm-snapshot-dialog/volume-snapshot-from-vm-snapshot-dialog.component.html [hunk 6]
```diff
@@ -1,5 +1,5 @@
 <form [formGroup]="form" (ngSubmit)="onSubmit()">
-  <h2 mat-dialog-title="">{{ 'SNAPSHOT_PAGE.MODALS.TAKE_A_SNAPSHOT' | translate }}</h2>
+  <h2 mat-dialog-title>{{ 'SNAPSHOT_PAGE.MODALS.TAKE_A_SNAPSHOT' | translate }}</h2>
```
Comment (wowshakhov) [inline src/app/shared/components/volume-snapshot-from-vm-snapshot-dialog/volume-snapshot-from-vm-snapshot-dialog.component.html]: I think "_A_" is redundant. "TAKE_SNAPSHOT" is enough

File: src/app/shared/services/vm-snapshot.service.ts [hunk 7]
```diff
@@ -43,10 +53,9 @@ export class VmSnapshotService extends BaseBackendService<VmSnapshot> {
       );
   }
 
-  public revert(id: string): Observable<any> {
-    // todo it returns vm, maybe we need to update vms
+  public revert(id: string): Observable<VirtualMachine> {
     return super
       .sendCommand('revertTo', { vmsnapshotid: id })
-      .pipe(switchMap(job => this.asyncJobService.queryJob(job, this.entity)));
+      .pipe(switchMap(job => this.asyncJobService.queryJob(job, 'virtualmachine')));
```
Comment (wowshakhov) [inline src/app/shared/services/vm-snapshot.service.ts]: Why do you use `this.entity` in `create` and `virtualmachine` in `revert`?

File: src/app/vm/components/vm-snapshot-list-dialog/vm-snapshot-list-dialog.component.html [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    <ng-container matColumnDef="date">
+      <th mat-header-cell *matHeaderCellDef>{{ 'VM_PAGE.MODALS.VM_SNAPSHOT_LIST.DATE' | translate }}</th>
+      <td mat-cell *matCellDef="let vmSnapshot">
+        {{ vmSnapshot.created | csTimestampToDate | csStringifyDate }}
+      </td>
+    </ng-container>
+    <ng-container matColumnDef="actions">
+      <th mat-header-cell *matHeaderCellDef>{{ 'VM_PAGE.MODALS.VM_SNAPSHOT_LIST.ACTIONS' | translate }}</th>
+      <td mat-cell *matCellDef="let vmSnapshot">
+        <cs-entity-action-row [actions]="(vmSnapshotActionsEntities$ | async)[vmSnapshot.id]"></cs-entity-action-row>
```
Comment (wowshakhov) [inline src/app/vm/components/vm-snapshot-list-dialog/vm-snapshot-list-dialog.component.html]: `[vmSnapshot.id]` can you do it in the component?

File: src/app/vm/components/vm-snapshots-sidebar-card/vm-snapshots-sidebar-card.component.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  selector: 'cs-vm-snapshots-card',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  templateUrl: './vm-snapshots-sidebar-card.component.html',
+  styleUrls: ['./vm-snapshots-sidebar-card.component.scss'],
+})
+export class VmSnapshotsSidebarCardComponent implements OnInit {
+  @Input()
+  public lastVmSnapshot: VmSnapshot;
+  @Input()
+  public vmSnapshotsCount: number;
```
Comment (wowshakhov) [inline src/app/vm/components/vm-snapshots-sidebar-card/vm-snapshots-sidebar-card.component.ts]: Can you just pass all snapshots? You can then get the number of snapshots and the last snapshot.

File: src/app/vm/components/vm-snapshots-sidebar-card/vm-snapshots-sidebar-card.component.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  @Input()
+  public entityActions: EntityAction[];
+  @Output()
+  public createButtonClicked = new EventEmitter<void>();
+  @Output()
+  public showAllVMSnapshotsButtonClicked = new EventEmitter<void>();
+
+  constructor() {}
+
+  ngOnInit() {}
```
Comment (wowshakhov) [inline src/app/vm/components/vm-snapshots-sidebar-card/vm-snapshots-sidebar-card.component.ts]: Remove empty constructor and ngOnInit

File: src/app/root-store/server-data/vm-snapshots/vm-snapshots.effects.ts (hunk 1 above)
Comment (tamazlykar) [reply src/app/root-store/server-data/vm-snapshots/vm-snapshots.effects.ts]: Forgot to remove IN

File: src/app/vm/vm-sidebar/vm-detail/vm-snapshot-creation-dialog/vm-snapshot-creation-dialog.component.html [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    <mat-form-field class="full-width-input">
+      <input matInput formControlName="name" [placeholder]="'COMMON.NAME' | translate">
+    </mat-form-field>
+    <mat-form-field class="full-width-input">
+      <input matInput formControlName="description" [placeholder]="'COMMON.DESCRIPTION' | translate">
+    </mat-form-field>
+    <mat-checkbox formControlName="snapshotMemory">
+      {{ 'VM_PAGE.MODALS.TAKE_A_VM_SNAPSHOT.SNAPSHOT_MEMORY' | translate }}
+    </mat-checkbox>
+    <div class="warning-info">{{ 'VM_PAGE.MODALS.TAKE_A_VM_SNAPSHOT.WARNING_INFORMATION' | translate }}</div>
```
Comment (wowshakhov) [inline src/app/vm/vm-sidebar/vm-detail/vm-snapshot-creation-dialog/vm-snapshot-creation-dialog.component.html]: Maybe remove "_A_"?

File: src/i18n/en.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (wowshakhov) [reply src/i18n/en.json]: Just fix the typo then

File: src/app/shared/services/vm-snapshot.service.ts (hunk 7 above)
Comment (tamazlykar) [reply src/app/shared/services/vm-snapshot.service.ts]: Async service resolves data coming from API based on the second parameter of `queryJob` function. Create action returns VM snapshot, but revert action return VM.

File: src/app/vm/components/vm-snapshots-sidebar-card/vm-snapshots-sidebar-card.component.ts (hunk 9 above)
Comment (tamazlykar) [reply src/app/vm/components/vm-snapshots-sidebar-card/vm-snapshots-sidebar-card.component.ts]: This is because I decided to extract logic from the component and make it dumb.

File: src/app/root-store/server-data/vm-snapshots/vm-snapshots-entity-actions.selectors.ts [hunk 12]
```diff
@@ -14,28 +14,29 @@ export const getVmSnapshotEntityActionsEntities = createSelector(
...
+          disabled: !isSnapshotReady,
           visible: false,
           actionCreator: () =>
             new vmSnapshotsActions.CreateVolumeSnapshot({ snapshotId: vmSnapshot.id }),
         },
         {
           icon: 'mdi-backup-restore',
           text: 'SNAPSHOT_PAGE.ACTIONS.REVERT_VM_TO_SNAPSHOT',
-          disabled: vmState !== VmState.Running,
+          disabled: !isSnapshotReady || vmState !== VmState.Running,
```
Comment (wowshakhov) [inline src/app/root-store/server-data/vm-snapshots/vm-snapshots-entity-actions.selectors.ts]: remove vmState !== VmState.Running

File: src/app/vm/vm-sidebar/vm-detail/vm-snapshot-creation-dialog/vm-snapshot-creation-dialog.component.html [hunk 13]
```diff
@@ -3,10 +3,26 @@ <h3 mat-dialog-title>{{ 'VM_PAGE.MODALS.TAKE_VM_SNAPSHOT.TITLE' | translate }}</
 
   <mat-dialog-content>
     <mat-form-field class="full-width-input">
-      <input matInput formControlName="name" [placeholder]="'COMMON.NAME' | translate">
+      <input
+        matInput
+        formControlName="name"
+        [placeholder]="'COMMON.NAME' | translate"
+        [maxlength]="255"
```
Comment (wowshakhov) [inline src/app/vm/vm-sidebar/vm-detail/vm-snapshot-creation-dialog/vm-snapshot-creation-dialog.component.html]: consider extracting 255 to a global variable

File: src/app/vm/vm-sidebar/vm-detail/vm-snapshot-creation-dialog/vm-snapshot-creation-dialog.component.ts [hunk 14]
```diff
@@ -24,4 +24,12 @@ export class VmSnapshotCreationDialogComponent {
   public onClose() {
     this.dialogRef.close();
   }
+
+  public getNameControl(): AbstractControl | null {
+    return this.form.get('name');
```
Comment (wowshakhov) [inline src/app/vm/vm-sidebar/vm-detail/vm-snapshot-creation-dialog/vm-snapshot-creation-dialog.component.ts]: consider replacing with
`name = this.form.get('name')`

In template:
```
{{ name.value }}
```

File: src/app/reducers/vm/redux/vm.reducers.ts [hunk 15]
```diff
@@ -107,7 +106,8 @@ export function listReducer(state = initialListState, action: vmActions.Actions)
       };
     }
 
-    case vmActions.VIRTUAL_MACHINE_LOADED: {
+    case vmActions.VIRTUAL_MACHINE_LOADED:
+    case VmSnapshotActionTypes.RevertSuccess: {
```
Comment (MaverickW) [inline src/app/reducers/vm/redux/vm.reducers.ts]: Strange and inconsistent action types naming. 

Also, you put an action related to another reducer here. I guess it shouldn't be done like that. Based on my knowledge of the redux pattern,  there should be special action related exactly for this reducer.

File: src/app/reducers/vm/redux/vm.reducers.ts [hunk 16]
```diff
@@ -1,27 +1,26 @@
...
-
+import { noGroup } from '../../../vm/vm-filter/vm-filter.component';
 import * as fromAccounts from '../../accounts/redux/accounts.reducers';
-import * as vmActions from './vm.actions';
-import * as fromSGroup from '../../security-groups/redux/sg.reducers';
 import * as affinityGroupActions from '../../affinity-groups/redux/affinity-groups.actions';
+import * as fromSGroup from '../../security-groups/redux/sg.reducers';
 import * as fromZones from '../../zones/redux/zones.reducers';
-import * as uniq from 'lodash/uniq';
+import * as vmActions from './vm.actions';
```
Comment (MaverickW) [inline src/app/reducers/vm/redux/vm.reducers.ts]: A lot of unnecessary updates in the imports section.

I guess you have a different auto-formatting setting for your IDE in comparison with your teammates.

File: src/app/root-store/server-data/vm-snapshots/vm-snapshots.state.ts [hunk 17]
```diff
@@ -1,12 +1,19 @@
 import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';
 import { VmSnapshot } from '../../../shared/models/vm-snapshot.model';
+import moment = require('moment');
```
Comment (MaverickW) [inline src/app/root-store/server-data/vm-snapshots/vm-snapshots.state.ts]: `require` is a more nodejs way to import dependency. I prefer to use `import * as moment from 'moment'`

Comment (MaverickW) [reply src/app/root-store/server-data/vm-snapshots/vm-snapshots.state.ts]: You increasing coupling with `moment` by injecting it directly into this class. For example, we prefer to encapsulate the logic related to dates inside of some kind of `DatesService` and use it inside of our code. It would allow us to change `moment` to any other library which works with dates without any additional changes in other parts.

File: src/app/shared/constants.ts [hunk 18]
```diff
@@ -0,0 +1,6 @@
+export class Constants {
+  /**
+   * Max length 255 for some entities fields
+   */
+  public static readonly maxLength255 = 255;
```
Comment (MaverickW) [inline src/app/shared/constants.ts]: I guess It's a bad constant naming example. Imagine that you will need to change `255` to `256` it will lead to changes in constant name, which leads to changes in other files. 

The key idea of such constants - is to reduce the number of places which requires changes in the case of constant value changes.

File: src/app/shared/services/vm-snapshot.service.ts [hunk 19]
```diff
@@ -1,17 +1,22 @@
...
 import { VmSnapshot } from '../models/vm-snapshot.model';
 import { AsyncJobService } from './async-job.service';
 import { BaseBackendService, CSCommands } from './base-backend.service';
 
+export const vmSnapshotEntityName = 'VMSnapshot';
+
 @Injectable()
 @BackendResource({
-  entity: 'VMSnapshot',
+  entity: vmSnapshotEntityName,
```
Comment (MaverickW) [inline src/app/shared/services/vm-snapshot.service.ts]: I guess constant usage is redundant here (except the situations when `vmSnapshotEntityName` reused somewhere).

File: src/app/shared/services/vm-snapshot.service.ts [hunk 20]
```diff
@@ -43,10 +56,9 @@ export class VmSnapshotService extends BaseBackendService<VmSnapshot> {
       );
   }
 
-  public revert(id: string): Observable<any> {
-    // todo it returns vm, maybe we need to update vms
+  public revert(id: string): Observable<VirtualMachine> {
     return super
       .sendCommand('revertTo', { vmsnapshotid: id })
-      .pipe(switchMap(job => this.asyncJobService.queryJob(job, this.entity)));
+      .pipe(switchMap(job => this.asyncJobService.queryJob(job, virtualMachineEntityName)));
```
Comment (MaverickW) [inline src/app/shared/services/vm-snapshot.service.ts]: It's strange to see `virtualMachineEntityName` here since it's related to the other service, which leads to broken encapsulation principles.

File: src/app/shared/shared.module.ts [hunk 21]
```diff
@@ -134,26 +150,17 @@ import { UserService } from './services/user.service';
...
-import {
-  RoundStateIndicatorComponent,
-  SquareStateIndicatorComponent,
-} from './components/state-indicator';
 import { UrlDirective } from './validators/directives';
-import { TimePickerComponent } from './components/time-picker/time-picker.component';
-import { DayPeriodComponent } from './components/day-period/day-period.component';
-import { DiskOfferingSelectorChooserComponent } from './components/disk-offering/disk-offering-selector/disk-offering-selector-chooser.component';
-import { DiskOfferingSelectorSliderComponent } from './components/disk-offering/disk-offering-selector/disk-offering-selector-slider.component';
-import { AngularDraggableModule } from 'angular2-draggable';
```
Comment (MaverickW) [inline src/app/shared/shared.module.ts]: A lot of unnecessary changes in imports.

File: src/app/shared/shared.module.ts [hunk 22]
```diff
@@ -111,6 +125,8 @@ import { HypervisorService } from './services/hypervisor.service';
 import { JobsNotificationService } from './services/jobs-notification.service';
 import { LocalStorageService } from './services/local-storage.service';
 import { LoginGuard } from './services/login-guard.service';
+// tslint:disable max-line-length
```
Comment (MaverickW) [inline src/app/shared/shared.module.ts]: What's the purpose of this suppression?

File: src/app/vm/container/vm-detail.container.ts [hunk 23]
```diff
@@ -69,8 +77,17 @@ export class VmDetailContainerComponent implements OnInit {
...
+  readonly lastVmSnapshotActions$ = this.lastVmSnapshot$.pipe(
+    filter(snapshot => !!snapshot),
+    switchMap(snapshot =>
+      this.store.pipe(select(vmSnapshotsSelectors.getVmSnapshotEntityActions(snapshot.id))),
+    ),
+  );
+  readonly vmSnapshotsCount$ = this.vmSnapshots$.pipe(map(vmSnapshots => vmSnapshots.length));
 
-  constructor(private store: Store<State>) {}
+  constructor(private store: Store<State>, public dialog: MatDialog) {}
```
Comment (MaverickW) [inline src/app/vm/container/vm-detail.container.ts]: Should `dialog: MatDialog` be public? It's unused in a template.
